tests/php/ORM/DataQueryTest.php

```diff
@@ -574,4 +578,235 @@ public function testExistsCreatesFunctionalQueries()
+        }
+
+        $dataQuery = new DataQuery($dataClass);
+        $dataQuery->with($name, $query, $cteFields, $onClause, $recursive);
+
+        foreach ($extraManipulations as $method => $args) {
+            $dataQuery->$method(...$args);
+        }
+
+        $expected = [];
+
+        if (isset($expectedItems['fixtures'])) {
+            foreach ($expectedItems['fixtures'] as $fixtureName) {
+                $expected[] = $this->idFromFixture($dataClass, $fixtureName);
+            }
+            $this->assertEquals($expected, $dataQuery->execute()->column('ID'));
+        }
+
+        if (isset($expectedItems['data'])) {
+            foreach ($expectedItems['data'] as $data) {
+                if (isset($data['fixtureName'])) {
+                    $data = $this->objFromFixture($dataClass, $data['fixtureName'])->toMap();
+                } else {
+                    $data['ClassName'] = null;
+                    $data['LastEdited'] = null;
+                    $data['Created'] = null;
+                    $data['Price'] = null;
+                    $data['ID'] = null;
+                }
+                $expected[] = $data;
+            }
+            $this->assertListEquals($expected, new ArrayList(iterator_to_array($dataQuery->execute(), true)));
+        }
+    }
+
+    /**
+     * tests the WITH clause, using a DataQuery as the CTE query
+     */
+    public function testWithUsingDataQuery()
+    {
+        $dataQuery = new DataQuery(SQLSelectTest\CteRecursiveObject::class);
+        $cteQuery = new DataQuery(SQLSelectTest\CteRecursiveObject::class);
+        $cteQuery->where([
+            '"SQLSelectTestCteRecursive"."ParentID" > 0',
+            '"SQLSelectTestCteRecursive"."Title" = ?' => 'child of child2'
+        ]);
+        $cteQuery->union(new SQLSelect(
+            '"SQLSelectTestCteRecursive"."ParentID"',
+            ['"hierarchy"', '"SQLSelectTestCteRecursive"'],
+            [
+                '"SQLSelectTestCteRecursive"."ParentID" > 0',
+                '"SQLSelectTestCteRecursive"."ID" = "hierarchy"."ParentID"'
+            ]
+        ));
+        $dataQuery->with('hierarchy', $cteQuery, ['ParentID'], ['ID' => 'ParentID'], true);
+
+        $expectedFixtures = [
+            'child2',
+            'parent',
+            'grandparent',
+        ];
+        $expectedData = [];
+        foreach ($expectedFixtures as $fixtureName) {
+            $expectedData[] = $this->objFromFixture(SQLSelectTest\CteRecursiveObject::class, $fixtureName)->toMap();
+        }
+        $this->assertListEquals($expectedData, new ArrayList(iterator_to_array($dataQuery->execute(), true)));
+    }
+
+    /**
+     * Tests that CTE queries have appropriate JOINs for subclass tables etc.
+     * If `$query->query()->` was replaced with `$query->query->` in DataQuery::with(), this test would throw an exception.
+     * @doesNotPerformAssertions
+     */
+    public function testWithUsingDataQueryAppliesRelations()
```

Comment:
Using `@doesNotPerformAssertions` is the correct way to mark that a test passes simply by not throwing an exception. There's another test in the codebase that also does this.
The alternative would be to either:
1. add an assertion in here like `$this->assertTrue(true);` which is redundant with the `@doesNotPerformAssertions` annotation, or
1. add an assertion testing the result of the query, which makes it look like we care about the result of the query, which we don't